### DISCUSSION TITLE
Adding comments directly to a PDF?

### DISCUSSION THREAD

**Discussion by drew-heider** [RyotaUshio/obsidian-pdf-plus#132] [Q&A]
Hi, thanks for making such a great plugin! I'm an academic and it's really resolved a lot of issues. I am curious if it is possible to add comments _directly_ to a PDF's margins, similar to how you can directly add highlights. I'm thinking something similar to the implementation in Adobe or FoxIt (pictured below) where you don't need a separate markdown document. I couldn't figure out how to do it, so wanted to see if it's even possible at all.

Thanks!

<img width="359" alt="Capture" src="https://github.com/RyotaUshio/obsidian-pdf-plus/assets/50001314/0778d003-b624-4384-9b78-4151229ee4ec">

**Comment by RyotaUshio** [comment]
I'm not sure if I understand you correctly but isn't this it?

https://github.com/RyotaUshio/obsidian-pdf-plus/assets/72342591/e824ce48-a605-4972-8978-e4fdf7007b0c

**Reply by samanthavenialogan** [reply]
Hey @RyotaUshio and @N3C2L 
I was stoked to find this video because I thought I'd found what I needed but it didn't go far enough--how can we copy that annotation AND the highlight in our copy templates into a block? 

I cannot find a way to ask the system to pull that comment annotation alongside the highlight in the code. 

what happens when I try to copy it all is that it collects the highlight but it doesn't include the annotation. I would like it to show as a block that includes the quote in italicized font, and then it includes my annotation below it, but both "text" and "selection" report the highlighted text - not the annotation. 

What is the Javascript variable I'm looking for in order to pull the annotation? 

-------------

What the notes look like: 
![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/3e1fd6c2-6674-4a87-a34d-debb10c42ef9)

How I've asked it to port: 
![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/ecf6c20e-a1f4-49fa-935e-6af8d3e3ed41)

How it shows up:
![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/5a17b685-afd5-4b8d-ab52-04feec9729e6)


I presume that there is an annotation-specific field I need to pull in order for it to show, but the variable library above doesn't seem to have one that pulls the annotation contents as opposed to the quote. What can I do?

**Reply by N3C2L** [reply]
Afaik, this isn't possible currently. There is no variable which you can use to extract the annotation to your markdown annotation file. 
See this answer from @RyotaUshio in this thread:
https://github.com/RyotaUshio/obsidian-pdf-plus/discussions/112#discussioncomment-9007179

**Reply by samanthavenialogan** [reply]
Hmm, thank you @N3C2L for this and @RyotaUshio for directly approaching the issue. I'm unsure if you'd like me to just resume conversation on that quoted thread but at least for my purposes, it's actually really obtrusive to have to add my annotations manually. Either I have to remember to do that instead of using that exact functionality in the PDF annotator (which is the point of that function in the first place), or I have to manually C/P it when I am going through and C/Ping it to a separate note after the fact, which defeats the point of a mass C/Per in the first place. 

Either way, providing the space to put the data in the original document is AWESOME so it's at least there, but not making it available goes against the 'simple data is best' philosophy. Allowing for it to be there but not movable is just annoying because that means it's not available, not simplified, and not usable as a result? 

I also understand this seems to be less an issue of the plugin and more an issue of the attribute availability not targeting it. I believe that's something the core has to provide instead of the plugin? That's definitely above my current knowledge level but that's how I presume the attribute production works so I don't blame the plugin for not having this functionality. 

just another reason Obsidian DESPERATELY needs this plugin to be wrapped into the core build. 

Regardless if I want this feature it leaves only 1 singular option: use Kindle's Reader instead of obsidian and that SUCKS. Kindle can pull this annotation, and then obsidian can pull form Kindle so that's how you'd automatically get it - but Kindle doesn't write direct to the PDF in the google drive file which means Kindle once-again has a feature that holds me hostage. (I Figging hat Kindle but it's still an industry requirement until Obsidian wraps this plugin into core IF that happens.)

**Reply by RyotaUshio** [reply]
We already have the `text` variable, which for the text annotated by a highlight, but as you pointed out, we don't have one for the annotation content/comment.
Of course I agree that we need another variable for that, but the problem is that the annotation comment is specific to annotations written directly into PDF files and not applicable to backlink highlighting, which makes it difficult to keep the variable system consistent between these two types of annotations.

But here is a good news: for the meantime, you can get the annotation comment by `pdfPlus.lib.getAnnotation(new URLSearchParams(linktext.split("#")[1]).get("annotation"))?.data?.contentsObj?.str ?? ""`

For example: 

```
> > {{text}}
> 
> {{pdfPlus.lib.getAnnotation(new URLSearchParams(linktext.split("#")[1]).get("annotation"))?.data?.contentsObj?.str ?? ""}}
```

<img width="649" alt="image" src="https://github.com/RyotaUshio/obsidian-pdf-plus/assets/72342591/6bb46491-4622-44d5-ad51-11555b3f21a2">

<img width="440" alt="image" src="https://github.com/RyotaUshio/obsidian-pdf-plus/assets/72342591/06c24fbd-de8b-4598-8a57-d914ee2afa65">

Some breakdown:

- `new URLSearchParams(linktext.split("#")[1]).get("annotation")` extracts the annotation id (e.g. `123R`) from the given link text (e.g. `file.pdf#page=1&annotation=123R`). You can also use some regexp instead.
- `pdfPlus.lib.getAnnotation(id)` fetches the information about the annotation with the given ID

If you want to make the code look clean or do more complex stuff, use Templater startup script to define global functions that can be used inside `{{...}}`. See https://github.com/RyotaUshio/obsidian-pdf-plus/discussions/145#discussioncomment-9084692 for more details and examples.

**Reply by samanthavenialogan** [reply]
Hello, 

Sadly this is kind of all greek to me, but because I don't necessarily have the knowledge to implement this. At the moment I'm good enough to use enough coding to do measurementmarketing tasks within Tag Manager for instance. That said, is this something that I have to run regularly on a manual basis or is it something that I can just set and develop the filepath as a variable? 

In short, I don't know where to go or where to even insert this code within obsidian, much less implement it using the global startup script. 

I am intimidated XD

**Reply by samanthavenialogan** [reply]
Back after asking a more savvy friend how this works -- we got it to work! There are a few other hangups that I have specifically having to do with some weird markup conversions between obsidian and Airtable but for the most part, this was very helpful! My friend new exactly what to do to make the code you provided work ^^

**Reply by RyotaUshio** [reply]
I've created a related issue: #225

**Reply by samanthavenialogan** [reply]
This issue has been resolved in the newest patch :) I've commented with resolution on the related issue linked above.

**Reply by samanthavenialogan** [reply]
Hey @RyotaUshio, 

I have sadly come across another PDF issue that is having the same problem (though weirdly better) again :/ 

I've uploaded the PDF I'm struggling with here for your review: 
[Paper-Antisocial Online Behavior Detection Using Deep Learning-community_moderation-zinovyeva-2019.pdf](https://github.com/user-attachments/files/16061103/Paper-Antisocial.Online.Behavior.Detection.Using.Deep.Learning-community_moderation-zinovyeva-2019.pdf)

Once again I have highlighted and commented throughout the document but the copy/paste feature shows empty comments again. 
![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/0586768c-de9a-4898-b347-5190bd3a3caf)

![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/87299509-fe98-4288-9c37-38011cb469d7)

The weird part is that it will pull the comments on highlights that are currently 'on screen' but it seems like the document is 'lazy loading' Those changes and comments a certain distance form the screen load will not pull. I can't figure out what the distance is before it stops rendering comments though. It will capture several pages +/- the window's load. If you scroll down to page 22 for instance the comment there will load but the beginning comments will not. If you go to page 9, it again, copies comments on that highlight but none of the others at the start of the document. 

HILARIOUSLY, I thought it would be an amazing workaround of 'just zoom out' and that absolutely improved the issue. I was able to get all highlights by zooming to 10% and then going to page 8 (half way) and it collected them all. I have not found this issue on any other PDFs so far and I'll hunt around for more as I pursue research but I'm concerned what will happen when I pull up longer documents. 

![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/76f8d95e-2876-4ad9-99a4-fc5d074122c5)

_________________
Critically, when using a specific highlight C/P, it does pull the comment. I have not changed my prompt for the comments changes. 

```
> [!{{calloutType}} ] {{linkWithDisplay}}
>>**Quote**: _{{selection}}_
>
>>**Comment**:  {{comment}}`
```

**Reply by samanthavenialogan** [reply]
Hey @RyotaUshio  just wanted to follow up on this? 

I does indeed look like it might  be a lazy load issue. Just attempted to pull a 50 page document and from 25 it pulled only from 13 to 33. 

Here is the document: 
[paper-The Rise of Content Cartels-politics-Evelyn Douek-2019.pdf](https://github.com/user-attachments/files/16124970/paper-The.Rise.of.Content.Cartels-politics-Evelyn.Douek-2019.pdf)

**Reply by samanthavenialogan** [reply]
@RyotaUshio  can I by chance get an update on this pdf viewer issue? it won't copy comments outside of the load menu on PDFs but when I zoom out to make sure as many pages as possible are there it will load more. Sadly I'm often reading entire books and Ph.D manuscripts for review that are over 100 pages and they cannot possibly fit on view windows beyond 40-60 pages.

**Comment by drew-heider** [comment]
Oops sorry, let me clarify! I mean is it possible to have an icon next to a highlight that contains an annotation to add a visual cue to where comments are?

**Comment by N3C2L** [comment]
You could add a new color and change the name of this color to "comment".
If your annotation contains a comment, right click and change the color to your "comment" color.
For example: All highlights that don't contain comments have a green highlight color.
And for highlights that contain a comment, you have a red highlight color.
That's easy to distinguish and you can even add more colors named "Questions" or "Research" etc.
You only need to learn the colors and you know what's "behind" a highlight.

**Reply by drew-heider** [reply]
Thanks for this! I've started using this system. The only issue is that I'm color blind so at times it can be a bit difficult to distinguish colors if there's more than one on a single page. It would be great if we could have another visual cue for comments as well, similar to how other pdf readers will show a little sticky note in the margins or box.

Thanks for your help with this!

**Reply by samanthavenialogan** [reply]
Isn't there a color blind obsidian theme? it's plausible you could implement that and then adapt the PDF++ to that system?

**Comment by Dennyyusuf** [comment]
Hello I don't know that this is possible. Can you please teach me on how to make a comment directly in the PDF without copy / paste the highlight into markdown file?

**Reply by samanthavenialogan** [reply]
Hey @Dennyyusuf, 

I implemented my setup like this: 
1. Ensure PDF++ is installed, enabled and fully updated. 
2. go into the plugin settings and ensure the following settings are set to 'on'
     1.  Enable PDF Editing | ON
     2. Annotation author [your name. an annotator should be mentioned for the files metadata.]
     3. Highlight backlinks in PDF viewer | ON
     4. Colors [confirm your highlights are properly named for their intended purpose so it comes through in the template]
     ![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/6045b8ea-07a9-4828-b73a-773fa0d708c1)
     5. Under menu items set all of them to ON except optional ones like read aloud selected text.
     6. set a hotkey for copying links. Once a hotkey is set, if you didn't select an annotation you can copy ALL of them by default. 
     [I set mine to ctrl+shift+C using the hotkey menu in obsidian.    
     7. there are TON of other optional features so I might recommend spending some time exploring too ^^ 
![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/b5fa771c-3378-4dd4-98bf-e2de56e7e2db)

7. Now under 'custom copy formats' Include a new 'quote in callout' option under "custom copy formats" section of PDF++
![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/4dc8225d-fb16-4e89-9bbf-c0751573f947)

This is the format I use which causes obsidian notes to render like this: 
```
> [!{{calloutType}} ] {{linkWithDisplay}}
>>**Quote**: _{{selection}}_
>
>>**Comment**:  {{comment}}
```
![image](https://github.com/RyotaUshio/obsidian-pdf-plus/assets/54729415/028c52ac-9395-40d0-81a7-942b9d20dc0a)

**Reply by Dennyyusuf** [reply]
Thank you @samanthavenialogan , for giving me thorough, step-by-step guidance. I appreciate the effort.

However, I wonder why, after annotating the PDF with comments, I still can't put the comments in the Markdown file. I already used your template. It just unable to copy the annotated text with the comment.

> [!{{calloutType}} ] {{linkWithDisplay}}
>>**Quote**: _{{selection}}_
>
>>**Comment**:  {{comment}}

**Reply by samanthavenialogan** [reply]
Hmm, it's plausible you're running into the same problem I was with the PDF that you're using and the PDF is actually not able to be pulled from, but we wouldn't be able to rule out simpler issues first to be sure that's the case. 

So starting with the basics (I know, frustrating) can you please provide some screenshots on what you are doing right now? Then tell us what you expect to happen, and lastly what is happening? We'd love screenshots, or even better a video if you're willing ^^ 

Additionally, if you could share the PDF I'm happy to give it a go on my side since we know that my implementation works. If I can pull them we can narrow it to the implementation, instead of mucking around with the PDF.

**Reply by Dennyyusuf** [reply]
Hello Samatha Thanks for your feedback. 
Yes I thought so after reading above discussion.
https://www.veed.io/view/6d1be3b1-7678-49a1-b9fb-a94d67167b75?panel=share

**Reply by samanthavenialogan** [reply]
@Dennyyusuf  WOW YOU HAVE A LOT OF PLUGINS XD Have you tried running PDF+ in the sandbox to make sure another plugin isn't interfering? I had another PDF reader installed a few months back and it was trying to do both systems at the same time - it  broke things that I discovered when @RyotaUshio was asking me to do sandbox. 

Thanks for the video and I'm sorry I didn't respond so long. I haven't visited GitHub in a LONG time, because I've been so waterlogged with the Ph.D. Is the problem still persisting at the moment?

Assuming that it is still a problem, I think the big issue is that your comment systems are not being pulled as a comment by default because you copied the 'quote in callout' for a single comment under the comments format (bottom right in the image posted) and your default is also on 'page only'  When you pasted the comment if followed the template for your page only listing instead of comment. 

The easiest way to see if this is the case is to set your global copy/paste to comment, and then copy the global one. If that copies in comment format then its likely your selection default that's doing it. Once you change that selecting default to match comment instead of page only it should follow the comment template instead of the page template. 

![image](https://github.com/user-attachments/assets/df20489d-b0fe-4dc4-8b67-db10b16790a3)
___________________________________________________________________________________Change this to comment_ ^

I will note that this is conjecture if I'm honest, and I am personally experiencing a potentially related issue that I hope @RyotaUshio  can fix in which copying all comments in a PDF will pull all over the highlights but it will not pull all of the comments unless the window actually has the comments loads as I've outlined in this issue: https://github.com/RyotaUshio/obsidian-pdf-plus/discussions/223

Effectively if the window isn't loaded the highlights come through but the comment does not. I haven't seen the behavior that you're suggesting but it may be worth opening the document in Adobe, scrolling to that page, and trying again with the PDF clearly loaded in Adobe and Obsidian and on the window displayed.
